fix(appstore-connect): Add a new endpoint that returns the status of all configured apps

relaxolotl opened the pull request:
WIP. starts work on a new validate endpoint that hopefully requires fewer round trips from the FE and includes additional data.

relaxolotl commented on static/app/types/debugFiles.tsx:
```diff
@@ -65,6 +65,30 @@ export type AppStoreConnectValidationData = {
   updateAlertMessage?: string;
 };
 
+export type AppStoreConnectStatusData = {
```
@priscilawebdev this probably has a bulk of what you're interested in.

i went for 
```appstoreCredentialsValid: {is_valid: boolean} | {error: string; detail: string};``` because i'm thinking that it's a little easier to work around type-wise for you instead of say ```{ status: string; detail: string | undefined }```. happy to change it to something else though.

i'm also thinking that `detail` actually is unnecessary for us to return since you can probably just derive a string to be shown to the user given some error value. the error's supposed to be an enum anyways but it's a WIP i'm just not sure what its range of values look like so far. have to get some test data set up first. back to `detail` though, i would imagine not including it can cut down on payload sizes and it's probably not too compatible with i18n anyways.

thoughts?

priscilawebdev replied:
oh... maybe we can just assume that if the backend returns an error then it's invalid... so it would be something like `{error: string | null}`? otherwise, I think I would still prefer` { status: string; detail: string | undefined }`

relaxolotl commented on src/sentry/api/endpoints/project_app_store_connect_credentials.py:
````diff
@@ -411,3 +411,119 @@ def get(self, request: Request, project: Project, credentials_id: str) -> Respon
+    ``GET projects/{org_slug}/{proj_slug}/appstoreconnect/status``
+
+    Response:
+    ```json
````
here's the proposed json response that's supposed to line up with the TS type.

relaxolotl commented:
i need to get a few test api keys set up to see what apple returns in a few easy-to-reproduce situations.

relaxolotl commented on src/sentry/api/endpoints/project_app_store_connect_credentials.py:
```diff
@@ -411,3 +415,128 @@ def get(self, request: Request, project: Project, credentials_id: str) -> Respon
+            except appstore_connect.UnauthorizedError:
+                asc_credentials = {
+                    "status": "invalid",
+                    "code": AppConnectAuthenticationError().code,
```
I plan to throw these errors directly in `/app` so the FE can use the same code to handle the validation messages returned by this endpoint as well as `/status`

flub replied:
I'm not sure you need to instantiate the class here, can you not access the code attribute without instantiating?

relaxolotl replied:
you're right about that, thanks for catching this!

relaxolotl commented on static/app/types/debugFiles.tsx:
```diff
@@ -1,3 +1,5 @@
+import {ValidationErrorDetailed} from 'app/components/modals/debugFileCustomRepository/appStoreConnect/utils';
```
this doesn't introduce a circular import but i'm not particularly confident about this import. maybe it's better to just duplicate the type here instead?

priscilawebdev replied:
should we move the `ValidationErrorDetailed` type and maybe add some `AppStoreConnect` prefix or rename it?

relaxolotl replied:
do you have any ideas on where `ValidationErrorDetailed` could be moved to? i'm not too familiar with the front end's file structure so hints here would be very helpful.

priscilawebdev replied:
please do not worry... I worked on this in the PR https://github.com/getsentry/sentry/pull/29843

relaxolotl commented on static/app/components/projects/appStoreConnectContext/utils.tsx:
```diff
@@ -1,17 +1,20 @@
-    'The credentials of your configured App Store Connect are invalid.'
-  ),
-};
+export function areAppStoreConnectCredentialsValid(
+  credentialsStatus: AppStoreConnectCredentialsStatus | undefined
+): boolean {
+  return credentialsStatus?.status === 'valid';
+}
```
i'm not sure if this belongs here, open to suggestions on better places to put this helper

relaxolotl commented:
@ potential reviewers: i would really recommend reviewing this commit-by-commit, or to at least review the code without 6f2e5a6 included as it performs a rename that only clutters the diff

flub commented on src/sentry/api/endpoints/project_app_store_connect_credentials.py:
```diff
@@ -472,7 +476,28 @@ def get(self, request: Request, project: Project) -> Response:
+            else:
+                if apps:
+                    asc_credentials = {"status": "valid"}
+                else:
```
Hum, Is this really true?  I think you're now returning invalid even when the credentials were fine but there were not apps to list, which isn't really true is it?  I think you already handle invalid above so lines 489-495 should all be considered valid should they not?

(i strongly suspect the appstore_connect.get_apps() docstring and signature is outdated and that it's `except ValueError` clause is never triggered anymore)

relaxolotl replied:
> Hum, Is this really true? 

i don't think it's necessarily true, but it is what the code has historically been doing when apps aren't available. i'm open to the idea of changing this to `Forbidden` as it's likely that this is will happen when the token hasn't been granted permissions to an app (or there are no apps at all..?), but it's not too clear to me what you're hoping to see at this line instead.

honestly, i don't really have the right tools and permissions to be setting up these test scenarios myself as i'm unable to create API tokens, so i'm really just working off of fumes here.

whatever change is made here, i'm pretty sure it'll be new behaviour relative to the code prior to this change. i'm a little hesitant to do that, but if you feel it's important to add it in to this PR i can do that.

flub replied:
apologies, I wasn't very clear here.  I'm concerned that this tells users their credentials are invalid when they have no apps.  I don't think we need to fix it now as it's a funny edge case and it probably is simpler to leave like this.  I've created a ticket to track this as future work though.

flub reviewed (approved):
Looks nice!

(but there are no tests :( )

relaxolotl commented:
fyi @priscilawebdev @flub i've rebased this off of `master` to resolve the conflicts and to prune some duplicate commits due to a recent merge of a dependency. i moved the rename commit back to the bottom of the commit stack to make it easier to omit it from the diff during review.

relaxolotl commented:
@flub working on tests now; will add those in a follow-up PR.

priscilawebdev commented on src/sentry/api/endpoints/project_app_store_connect_credentials.py:
````diff
@@ -411,3 +415,122 @@ def get(self, request: Request, project: Project, credentials_id: str) -> Respon
+    ```json
+    [
+        {
+            "id": "abc123",
````
since we get the id as a property, do we need the id inside of the object too?

ex:

```
{
 "4e1c91f46aea4101b2953192d61cef4b": {
      "id":"4e1c91f46aea4101b2953192d61cef4b",
      "credentials": ...
    }
}
 
```

relaxolotl replied:
good point, i'll see what i can do to remove that.

relaxolotl replied:
done 🎉

priscilawebdev commented on src/sentry/api/urls.py:
```diff
@@ -2018,6 +2019,11 @@
                     name="sentry-api-0-project-appstoreconnect-credentials-update",
                 ),
+                url(
+                    r"^(?P<organization_slug>[^\/]+)/(?P<project_slug>[^\/]+)/appstoreconnect/status$",
```
```suggestion
                    r"^(?P<organization_slug>[^\/]+)/(?P<project_slug>[^\/]+)/appstoreconnect/status/$",
```
🤷‍♀️

priscilawebdev commented on static/app/components/modals/debugFileCustomRepository/appStoreConnect/index.tsx:
```diff
@@ -114,6 +113,7 @@ function AppStoreConnect({
       const appStoreConnnectError = getAppStoreErrorMessage(error);
       if (typeof appStoreConnnectError === 'string') {
         // app-connect-authentication-error
+        // 'app-connect-forbidden-error'
```
```suggestion
        // app-connect-forbidden-error
```

priscilawebdev commented on static/app/components/modals/debugFileCustomRepository/appStoreConnect/index.tsx:
```diff
@@ -238,15 +238,27 @@ function AppStoreConnect({
-        </StyledAlert>
-      );
+    if (credentials?.status === 'invalid') {
+      switch (credentials.code) {
+        case 'app-connect-forbidden-error':
+          alerts.push(
+            <StyledAlert type="warning" icon={<IconWarning />}>
+              {t(
+                'Your App Store Connect credentials have insufficient permissions. To reconnect, update your credentials.'
+              )}
+            </StyledAlert>
+          );
+          break;
+        case 'app-connect-authentication-error':
+        default:
+          alerts.push(
+            <StyledAlert type="warning" icon={<IconWarning />}>
+              {t(
+                'Your App Store Connect credentials are invalid. To reconnect, update your credentials.'
+              )}
+            </StyledAlert>
+          );
+      }
```
```suggestion
 alerts.push(
        <StyledAlert type="warning" icon={<IconWarning />}>
          {credentials.code === 'app-connect-forbidden-error'
            ? t(
                'Your App Store Connect credentials have insufficient permissions. To reconnect, update your credentials.'
              )
            : t(
                'Your App Store Connect credentials are invalid. To reconnect, update your credentials.'
              )}
        </StyledAlert>
      );
```

priscilawebdev commented on static/app/components/modals/debugFileCustomRepository/appStoreConnect/utils.tsx:
```diff
@@ -116,3 +106,23 @@ export function getAppStoreErrorMessage(
+  switch (error.code) {
+    case 'app-connect-authentication-error':
+      return t(
+        'Credentials are invalid, missing, or expired. Check the entered App Store Connect credentials are correct and have not expired.'
```
```suggestion
'Credentials are invalid, missing, or expired. Check that the credentials are correct and that they have not expired.'
```

priscilawebdev replied:
Is it possible for an app store's credentials to expire? 🧐

relaxolotl replied:
good point: i was thinking it _could_ refer to credentials that have been redacted, but expired probably implies a temporal aspect to these that really isn't valid here. i'll remove the bits talking about expiration.

priscilawebdev commented on static/app/components/projects/appStoreConnectContext/utils.tsx:
```diff
@@ -1,17 +1,20 @@
-};
+export function areAppStoreConnectCredentialsValid(
+  credentialsStatus: AppStoreConnectCredentialsStatus | undefined
+): boolean {
```
```suggestion
) {
```

ts will automatically infer the return type here 🙂

priscilawebdev commented on static/app/views/settings/projectDebugFiles/externalSources/customRepositories/status.tsx:
```diff
@@ -23,9 +24,9 @@ function Status({details, onEditRepository}: Props) {
+  const {pendingDownloads, credentials, lastCheckedBuilds} = details ?? {};
 
-  if (appstoreCredentialsValid === false) {
+  if (areAppStoreConnectCredentialsValid(credentials) === false) {
```
```suggestion
  if (!areAppStoreConnectCredentialsValid(credentials)) {
```

priscilawebdev replied:
should we maybe just 

```suggestion
  if (credentials?.status !== 'valid') {
```

priscilawebdev commented on static/app/components/projects/appStoreConnectContext/utils.tsx:
```diff
@@ -1,17 +1,20 @@
-  if (appConnectValidationData.appstoreCredentialsValid === false) {
-    return appStoreConnectAlertMessage.appStoreCredentialsInvalid;
+  credentialsStatus: AppStoreConnectCredentialsStatus
+): string | undefined {
```
```suggestion
) {
```
ts will automatically infer the return type here 🙂

priscilawebdev commented on static/app/views/settings/project/appStoreConnectContext.tsx:
```diff
@@ -33,28 +33,30 @@ const Provider = withApi(
-        const response = await api.requestPromise(
-          `/projects/${orgSlug}/${project.slug}/appstoreconnect/validate/${appStoreConnectSymbolSourceId}/`
+        const response: Map<string, AppStoreConnectStatusData> = await api.requestPromise(
+          `/projects/${orgSlug}/${project.slug}/appstoreconnect/status`
```
```suggestion
          `/projects/${orgSlug}/${project.slug}/appstoreconnect/status/`
```

priscilawebdev commented on static/app/components/projects/appStoreConnectContext/index.tsx:
```diff
@@ -71,13 +71,15 @@ const Provider = ({children, project, organization}: ProviderProps) => {
-      const response = await api.requestPromise(
-        `/projects/${orgSlug}/${projectDetails.slug}/appstoreconnect/validate/${appStoreConnectSymbolSourceId}/`
+      const response: Map<string, AppStoreConnectStatusData> = await api.requestPromise(
+        `/projects/${orgSlug}/${projectDetails.slug}/appstoreconnect/status`
```
```suggestion
        `/projects/${orgSlug}/${projectDetails.slug}/appstoreconnect/status/`
```

priscilawebdev commented on static/app/components/projects/appStoreConnectContext/utils.tsx:
```diff
@@ -1,17 +1,20 @@
-  ),
-};
+export function areAppStoreConnectCredentialsValid(
+  credentialsStatus: AppStoreConnectCredentialsStatus | undefined
```
```suggestion
  credentialsStatus?: AppStoreConnectCredentialsStatus
```

priscilawebdev commented on static/app/views/settings/project/appStoreConnectContext.tsx:
```diff
@@ -2,11 +2,11 @@ import {createContext, useEffect, useState} from 'react';
 
```
can we pls delete this file? looks like it's not being used

relaxolotl replied:
if this gets removed i think this leaves `static/app/components/projects/appStoreConnectContext/utils.tsx` in a strange place. where do you suppose these files should sit instead? would `static/app/components/modals/debugFileCustomRepository/appStoreConnect/utils.tsx` be a good new home?

priscilawebdev replied:
we have 2 appStoreConnectContext... one is the file `'app/components/projects/appStoreConnectContext'` which we use in many different places and the other one is `static/app/components/projects/appStoreConnectContext` which apparently is not being used... but please do not worry... the used file will be removed in this PR https://github.com/getsentry/sentry/pull/29843/ 😉

relaxolotl commented:
status endpoint is currently throwing 405s for some reason, will investigate later after lunch

relaxolotl commented:
@priscilawebdev i believe i've addressed all of your feedback, so this should be good for a second look. if you're unable to find time to take a second look by tomorrow, i'll merge this in first and let's follow up on any remaining feedback you've got in another PR.

priscilawebdev reviewed (approved):
looks good to me

priscilawebdev commented on static/app/components/modals/debugFileCustomRepository/appStoreConnect/utils.tsx:
```diff
@@ -116,3 +106,23 @@ export function getAppStoreErrorMessage(
+  switch (error.code) {
+    case 'app-connect-authentication-error':
+      return t(
+        'Credentials are invalid or missing. Check the entered App Store Connect credentials are correct.'
```
<3
